pkg/controller/pxc/users.go

```diff
@@ -0,0 +1,330 @@
+			if errU != nil {
+				return errors.Wrap(errU, "update secret status")
+			}
+			return errors.Wrap(err, "manage users")
+		}
+
+		if cr.Spec.ProxySQL != nil && cr.Spec.ProxySQL.Size > 0 {
+			pod := corev1.Pod{}
+			err = r.client.Get(context.TODO(),
+				types.NamespacedName{
+					Namespace: cr.Namespace,
+					Name:      cr.Name + "-proxysql-0",
+				},
+				&pod,
+			)
+			if err != nil {
+				return errors.Wrap(err, "get proxy pod")
+			}
+			var errb, outb bytes.Buffer
+			err = r.clientcmd.Exec(&pod, "proxysql", []string{"proxysql-admin", "--syncusers"}, nil, &outb, &errb, false)
+			if err != nil {
+				return errors.Errorf("exec syncusers: %v / %s / %s", err, outb.String(), errb.String())
+			}
+			if len(errb.Bytes()) > 0 {
+				return errors.New("syncusers: " + errb.String())
+			}
+			log.Info(outb.String())
+		}
+
+		for key, pass := range usersSecretObj.Data {
+			if key == "grants.yaml" {
+				continue
+			}
+			usersSecretObj.Annotations[key] = sha256Hash(pass)
+		}
+		usersSecretObj.Annotations["last-applied"] = newHash
+		usersSecretObj.Annotations["status"] = statusSucceeded
+
+		err = r.updateInternalSecret(cr, &usersSecretObj)
+		if err != nil {
+			return errors.Wrap(err, "update internal secret")
+		}
+
+		err = r.client.Update(context.TODO(), &usersSecretObj)
+		if err != nil {
+			return errors.Wrap(err, "update secret status")
+		}
+		return nil
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) getInternalSecret(cr *api.PerconaXtraDBCluster, userSecret *corev1.Secret) (corev1.Secret, error) {
+	secretName := internalPrefix + cr.Spec.SecretsName
+	secretObj := corev1.Secret{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      secretName,
+		},
+		&secretObj,
+	)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return secretObj, errors.Wrap(err, "get internal users secret")
+	} else if k8serrors.IsNotFound(err) {
+		secretObj, err = r.handleUsers(cr, userSecret, false)
+		if err != nil {
+			return secretObj, errors.Wrap(err, "handle internal secret")
+		}
+
+		err = r.client.Create(context.TODO(), &secretObj)
+		if err != nil {
+			return secretObj, errors.Wrap(err, "create internal users secret")
+		}
+		return secretObj, nil
+	}
+
+	return secretObj, nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) updateInternalSecret(cr *api.PerconaXtraDBCluster, userSecret *corev1.Secret) error {
+	secretObj, err := r.handleUsers(cr, userSecret, true)
+	if err != nil {
+		return errors.Wrap(err, "update internal secret")
+	}
+	err = r.client.Update(context.TODO(), &secretObj)
+	if err != nil {
+		return errors.Wrap(err, "create internal users secret")
+	}
+	return nil
+
+}
+
+func usersDataChanged(newHash *string, usersSecret *corev1.Secret) bool {
+	if secretData, ok := usersSecret.Data["grants.yaml"]; ok {
+		hash := sha256Hash(secretData)
+		*newHash = hash
+		if lastAppliedHash, ok := usersSecret.Annotations["last-applied"]; ok {
```

Comment:
can we remove this `if` and rename function into something like `getSecretHash`?
goal: improve readability in `reconcileUsers` func

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/392/commits/6152288c13c92aa90e99f0c829dfa64f325e87c4